pep-9999.rst

```diff
@@ -81,9 +81,201 @@ includes stubs for Python's standard library and several third-party
+
+    def ip_func() -> None: ...
+
+
+    class Foo:  # leave only one empty line above
+        x: int
+    class MyError(Exception): ...  # leave an empty line after the class
+
+Module Level Dunder Names
+-------------------------
+
+Do not include ``__all__`` in type stubs.
+
+Module Level Attributes
+-----------------------
+
+Do not use an assignment for module-level attributes.
+
+Yes::
+
+    CONST: str
+    x: int
+
+No::
+
+    CONST = ""
+    x: int = 0
+    y: float = ...
+    z = 0  # type: int
+
+Classes
+-------
+
+Do not include docstrings in class bodies.
+
+Classes without bodies should use the ellipsis operator ``...`` in place
+of the body on the same line as the class definition.
+
+Yes::
+
+    class MyError(Exception): ...
+
+No::
+
+    class MyError(Exception):
+        ...
+    class AnotherError(Exception): pass
+
+Functions and Methods
+---------------------
+
+The bodies of functions and methods should consist of only the ellipsis
+operator ``...`` on the same line as the closing parenthesis and colon.
+Do not include docstrings.
+
+Yes::
+
+    def to_int1(x: str) -> int: ...
+    def to_int2(
+        x: str,
+    ) -> int: ...
+
+No::
+
+    def to_int1(x: str) -> int:
+        return int(x)
+    def to_int2(x: str) -> int:
+        ...
+    def to_int3(x: str) -> int: pass
+
+Use the ellipsis operator ``...`` in place of actual default argument
+values. Use an explicit ``Optional`` annotation instead of
+a ``None`` default.
+
+Yes::
+
+    def foo(x: int = ...) -> None: ...
+    def bar(y: Optional[str] = ...) -> None: ...
+
+No::
+
+    def foo(x: int = 0) -> None: ...
+    def bar(y: str = None) -> None: ...
+    def baz(z: Optional[str] = None) -> None: ...
+
+Private Classes
+---------------
+
+If a function or method returns an instance of a private class, the
+type stub should provide a public, stub-only class::
+
+    # Only available during type checking
+    class SomeData:
+        x: int
+        def foo() -> None: ...
+
+    def get_some_data() -> SomeData: ...
+
+Language Features
+-----------------
+
+Use the latest language features available as outlined
+in the Syntax_ section, even for stubs targeting older Python versions.
+Do not use quotes around forward references and do not use ``__future__``
+imports.
+
+Yes::
+
+    class Py35Class:
+        x: int
+        forward_reference: OtherClass
+    class OtherClass: ...
+
+No::
+
+    class Py35Class:
+        x = 0  # type: int
+        forward_reference: 'OtherClass'
+    class OtherClass: ...
+
+Types
+-----
+
+Use ``float`` instead of ``Union[int, float]``. For argument types,
+use ``bytes`` instead of ``Union[bytes, memoryview, bytearray]``.
+
+Use ``Text`` in mixed Python 2 and 3 stubs when something accepts both
+``str`` and ``unicode``. Avoid using ``Text`` in stubs for Python 2 or
+3 only or in branches supporting only one Python version.
+
+Yes::
+
+    if sys.version_info < (3,):
+        def foo(s: unicode) -> None: ...
```

Comment:
It means `Union[bytes, unicode]`. I just double checked that the following code type checks ok in Python 2 mode with mypy:

```python
def foo(x):
    # type: (unicode) -> None
    pass

foo("")
```

Reply:
Thanks for checking! Then I think we should be more explicit about this guideline - if it's what we want to recommend. Pytype doesn't do this promotion because we got repeated complaints about inability to specify `unicode`-only parameters. (Funnily enough, no one's ever said anything about `float`/`int`). One possible alternative (which is what the Google style guide currently recommends) for Python 2 code is `bytes` or `unicode` for that specific type, `Text` for a union of unicode and native str, and `Union[bytes, unicode]` for a union of unicode and bytes. I'm not happy that there's no enforcement mechanism for distinguishing between the second and third cases, but it seems to more or less work.

Reply:
Makes sense. I changed it accordingly.